Apps/TicketManagement/Services/RedisTicketService.cs

```diff
@@ -138,59 +132,36 @@ public async Task<Ticket> CheckIn(CheckInRequest checkInRequest)
         {
             Stopwatch stopwatch = new();
             stopwatch.Start();
-            Ticket? ticket;
             IDatabase database = this.connectionMultiplexer.GetDatabase();
             RedisValue redisTicket = await database.StringGetAsync($"{checkInRequest.Room}:{checkInRequest.Id}").ConfigureAwait(true);
-            if (redisTicket.HasValue)
-            {
-                ticket = JsonSerializer.Deserialize<Ticket>(redisTicket.ToString());
-                if (ticket != null && ticket.Nonce == checkInRequest.Nonce)
-                {
-                    if (ticket.CheckInAfter > this.dateTimeDelegate.UtcUnixTime)
-                    {
-                        ticket.Status = TicketStatus.TooEarly;
-                    }
-                    else
-                    {
-                        RoomConfiguration? roomConfig = this.GetRoomConfiguration(ticket.Room);
-                        bool admit = false;
-                        string member = ticket.Id.ToString();
-                        if (ticket.Status == TicketStatus.Queued)
-                        {
-                            (long participantCount, _, long position) = await this.RoomCounts(roomConfig, member).ConfigureAwait(true);
-                            ticket.QueuePosition = position + 1;
-                            admit = participantCount + position < roomConfig.ParticipantLimit;
-                        }
+            TicketCheckin.ValidateRedisTicket(redisTicket);
 
-                        ITransaction trans = database.CreateTransaction();
-                        if (admit)
-                        {
-                            ticket.Status = TicketStatus.Processed;
-                            ticket.QueuePosition = 0;
-                            _ = trans.SortedSetRemoveAsync(GetRoomName(roomConfig, WaitingKey), member);
-                            _ = trans.HashSetAsync(GetRoomName(roomConfig, ParticipantsKey), member, string.Empty);
-                        }
+            Ticket? ticket = JsonSerializer.Deserialize<Ticket>(redisTicket.ToString());
+            TicketCheckin.ValidateTicket(ticket, checkInRequest.Nonce, this.dateTimeDelegate.UtcUnixTime);
 
-                        this.CheckIn(trans, roomConfig, ticket);
-                        await trans.ExecuteAsync().ConfigureAwait(true);
-                    }
-                }
-                else
-                {
-                    ticket = new()
-                    {
-                        Status = TicketStatus.NotFound,
-                    };
-                }
+            // Using null forgiving operator on ticket as TicketCheckin.ValidateTicket will validate if ticket is not null.
+            RoomConfiguration? roomConfig = this.GetRoomConfiguration(ticket!.Room);
```

Comment:
This may be able to be avoided by adding a [NotNull] attribute to the ticket parameter in TicketCheckin.ValidateTicket, letting callers know that the parameter won't be null if the method returns.  For more info, see https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis.